base/array.jl

```diff
@@ -517,8 +517,25 @@ julia> collect(Float64, 1:2:5)
 """
 collect(::Type{T}, itr) where {T} = _collect(T, itr, IteratorSize(itr))
 
-_collect(::Type{T}, itr, isz::HasLength) where {T} = copyto!(Vector{T}(undef, Int(length(itr)::Integer)), itr)
-_collect(::Type{T}, itr, isz::HasShape) where {T}  = copyto!(similar(Array{T}, axes(itr)), itr)
+function copyto_check_length!(dest::Array, src)
+    len = length(dest)
+    i = 0
+    for x in src
+        i == len &&
```

Comment:
It looks like `copyto!` already does this check.

Reply:
Yes but it doesn't do the one below, and it cannot be done with access to `i`.

Reply:
Ok, then we should add a `_copyto_impl!` that takes a `strict::Bool` argument, using the current code from `copyto!`. Then it can be shared by both this code (passing `true` for strict) and `copyto!` (passing `false`).

Reply:
I can do that. Though note that the error message will be less specific (like the current one in `copyto!`), since `copyto!` cannot assume that the problem comes from an invalid declared length. So I'm not sure it's really better.

Reply:
Ok, either way. This does need to use `eachindex` though.

Reply:
Not if we keep this method, as it only applies to `Array`.

Reply:
Actually it can also be called on `OffsetArray` and possibly on other arrays in corner cases. So I've taken the `_copy_impl!` approach.